File: server/prompt-builder.js
/* =========================================
    プロンプト構造化システム
   ========================================= */

/*

## 概要
LLMリクエスト用のプロンプトを動的に構築するモジュール。
システムプロンプト、カスタムプロンプト、会話履歴、コンテキスト情報を統合。

## 主要機能
- SYSTEM_PROMPTS: ベースプロンプト、ルーター、統合指示を定義
- CONTEXT_TEMPLATES: コンテキスト情報テンプレート
- PromptBuilder: プロンプト構築ロジック（複数LLMプロバイダー対応）
- createPromptBuilder(): ファクトリー関数
- logPromptDebugInfo(): デバッグ用ログ出力

## 対応プロバイダー
Claude, OpenAI, Gemini, Local LLM の異なるプロンプト形式に対応
*/

// システムプロンプト定義（空：各エージェントが独自プロンプトを持つ）
export const SYSTEM_PROMPTS = {};

// コンテキスト情報テンプレート
export const CONTEXT_TEMPLATES = {
    basic: `[コンテキスト情報]
現在のディレクトリ: {{CURRENT_PATH}}
ファイル数: {{FILE_COUNT}}
現在編集中: {{CURRENT_FILE}}`,

    withCustomPrompt: `[コンテキスト情報]
現在のディレクトリ: {{CURRENT_PATH}}
ファイル数: {{FILE_COUNT}}
現在編集中: {{CURRENT_FILE}}
カスタムプロンプト: {{CUSTOM_PROMPT_NAME}} (有効)`,

    withOpenFile: `[コンテキスト情報]
現在のディレクトリ: {{CURRENT_PATH}}
ファイル数: {{FILE_COUNT}}
現在編集中: {{CURRENT_FILE}}

[現在開いているファイルの詳細]
{{OPEN_FILE_INFO}}`
};

// プロンプトビルダークラス
export class PromptBuilder {
    constructor() {
        this.systemPrompt = '';
        this.userMessage = '';
        this.context = {};
        this.conversationHistory = [];
        this.customPrompt = null;
        this.maxHistoryItems = 15;  // デフォルトの履歴数
    }

    // システムプロンプトを設定
    setSystemPrompt(promptKey) {
        if (SYSTEM_PROMPTS[promptKey]) {
            this.systemPrompt = SYSTEM_PROMPTS[promptKey];
        } else {
            throw new Error(`Unknown system prompt: ${promptKey}`);
        }
        return this;
    }

    // カスタムプロンプトを設定
    setCustomPrompt(customPrompt) {
        this.customPrompt = customPrompt;
        return this;
    }

    // コンテキスト情報を設定
    setContext(context) {
        this.context = context;
        return this;
    }

    // 会話履歴を設定
    setConversationHistory(history) {
        this.conversationHistory = history || [];
        return this;
    }

    // 履歴制限を設定
    setMaxHistoryItems(max) {
        this.maxHistoryItems = max;
        return this;
    }

    // ユーザーメッセージを設定
    setUserMessage(message) {
        this.userMessage = message;
        return this;
    }

    // テンプレート変数を置換
    _replaceTemplateVars(text, context) {
        const fileList = context.fileList || [];
        const conversationHistory = context.conversationHistory || [];
        
        return text
            .replace('{{CURRENT_PATH}}', context.currentPath || '/workspace')
            .replace('{{FILE_COUNT}}', fileList.length)
            .replace('{{FILE_LIST}}', JSON.stringify(fileList, null, 2))
            .replace('{{HISTORY_COUNT}}', conversationHistory.length)
            .replace('{{CURRENT_FILE}}', context.currentFile || 'なし')
            .replace('{{CUSTOM_PROMPT_NAME}}', this.customPrompt?.name || '')
            .replace('{{OPEN_FILE_INFO}}', context.openFileInfo || '');
    }

    // システムプロンプトを構築（シングルプロンプト化）
    buildSystemPrompt() {
        // カスタムプロンプトがある場合はそれだけを使用
        if (this.customPrompt && this.customPrompt.content) {
            return this.customPrompt.content;
        }

        // カスタムプロンプトがない場合はベースプロンプト + コンテキスト情報
        let prompt = this.systemPrompt;

        // コンテキスト情報を追加
        if (SYSTEM_PROMPTS.contextTemplate) {
            prompt += '\n\n' + SYSTEM_PROMPTS.contextTemplate;
        }

        // テンプレート変数を置換
        return this._replaceTemplateVars(prompt, this.context);
    }

    // コンテキスト情報を構築
    buildContextInfo() {
        let template;
        
        if (this.context.openFileInfo) {
            template = CONTEXT_TEMPLATES.withOpenFile;
        } else if (this.customPrompt && this.customPrompt.enabled) {
            template = CONTEXT_TEMPLATES.withCustomPrompt;
        } else {
            template = CONTEXT_TEMPLATES.basic;
        }
        
        return this._replaceTemplateVars(template, this.context);
    }

    // 会話履歴を構築（プロバイダー別）
    buildConversationHistory(provider) {
        const recentHistory = this.conversationHistory.slice(-this.maxHistoryItems);
        const messages = [];
        
        switch (provider) {
            case 'claude':
                for (const exchange of recentHistory) {
                    messages.push({ role: 'user', content: exchange.user });
                    if (exchange.ai) {
                        messages.push({ role: 'assistant', content: exchange.ai });
                    }
                }
                break;
                
            case 'openai':
            case 'local':
                for (const exchange of recentHistory) {
                    messages.push({ role: 'user', content: exchange.user });
                    if (exchange.ai) {
                        messages.push({ role: 'assistant', content: exchange.ai });
                    }
                }
                break;
                
            case 'gemini':
                // Gemini用の履歴形式を改善
                if (recentHistory.length > 0) {
                    let historyText = '';
                    for (const exchange of recentHistory) {
                        historyText += `【ユーザー】\n${exchange.user}\n\n`;
                        if (exchange.ai) {
                            // JSONレスポンスの場合は簡略表示
                            const aiResponse = exchange.ai;
                            try {
                                const jsonMatch = aiResponse.match(/\{[\s\S]*\}/);
                                if (jsonMatch) {
                                    const parsed = JSON.parse(jsonMatch[0]);
                                    historyText += `【アシスタント】\n${parsed.message || aiResponse}\n\n`;
                                } else {
                                    historyText += `【アシスタント】\n${aiResponse}\n\n`;
                                }
                            } catch (e) {
                                historyText += `【アシスタント】\n${aiResponse}\n\n`;
                            }
                        }
                    }
                    return historyText;
                }
                return '';
        }
        
        return messages;
    }

    // 最終的なユーザーメッセージを構築
    buildFinalUserMessage() {
        const contextInfo = this.buildContextInfo();
        return `${this.userMessage}\n\n${contextInfo}`;
    }

    // プロバイダー別の完全なプロンプトを構築
    buildForProvider(provider) {
        const systemPrompt = this.buildSystemPrompt();
        
        switch (provider) {
            case 'claude':
                const claudeMessages = this.buildConversationHistory('claude');
                claudeMessages.push({
                    role: 'user',
                    content: this.buildFinalUserMessage()
                });
                return {
                    system: systemPrompt,
                    messages: claudeMessages
                };
                
            case 'openai':
            case 'local':
                const messages = [{ role: 'system', content: systemPrompt }];
                const history = this.buildConversationHistory(provider);
                messages.push(...history);
                messages.push({
                    role: 'user',
                    content: this.buildFinalUserMessage()
                });
                return { messages };
                
            case 'gemini':
                const historyText = this.buildConversationHistory('gemini');
                const userMessage = this.buildFinalUserMessage();
                
                // Gemini向けにプロンプト構造を明確化
                const fullContent = 
                    "【システムプロンプト】\n" + systemPrompt + "\n\n" + 
                    "【過去の会話】\n" + historyText + 
                    "【現在の質問】\n【ユーザー】\n" + userMessage;
                return { content: fullContent };
                
            default:
                throw new Error(`Unsupported provider: ${provider}`);
        }
    }
}

// ファクトリー関数
export function createPromptBuilder() {
    return new PromptBuilder();
}

// デバッグ用ログ出力
export function logPromptDebugInfo(promptData, provider) {
    if (true) { // デバッグログを強制的に有効化
        console.log(`--- ${provider.toUpperCase()} API Request ---`);
        console.log('Prompt Data:');
        console.log(JSON.stringify(promptData, null, 2).replace(/\\n/g, '\n'));
        console.log('--------------------------');
    }
}

File: server_pythonver/ai/prompt_builder.py
import json

# システムプロンプト定義（空：各エージェントが独自プロンプトを持つ）
SYSTEM_PROMPTS = {}

# コンテキスト情報テンプレート
CONTEXT_TEMPLATES = {
    "basic": "[コンテキスト情報]\n現在のディレクトリ: {{CURRENT_PATH}}\nファイル数: {{FILE_COUNT}}\n現在編集中: {{CURRENT_FILE}}",

    "withCustomPrompt": "[コンテキスト情報]\n現在のディレクトリ: {{CURRENT_PATH}}\nファイル数: {{FILE_COUNT}}\n現在編集中: {{CURRENT_FILE}}\nカスタムプロンプト: {{CUSTOM_PROMPT_NAME}} (有効)",

    "withOpenFile": "[コンテキスト情報]\n現在のディレクトリ: {{CURRENT_PATH}}\nファイル数: {{FILE_COUNT}}\n現在編集中: {{CURRENT_FILE}}\n\n[現在開いているファイルの詳細]\n{{OPEN_FILE_INFO}}"
}

class PromptBuilder:
    def __init__(self):
        self.system_prompt = ""
        self.user_message = ""
        self.context = {}
        self.conversation_history = []
        self.custom_prompt = None
        self.max_history_items = 15  # デフォルトの履歴数

    def set_system_prompt(self, prompt_key: str):
        if prompt_key in SYSTEM_PROMPTS:
            self.system_prompt = SYSTEM_PROMPTS[prompt_key]
        else:
            raise ValueError(f"Unknown system prompt: {prompt_key}")
        return self

    def set_custom_prompt(self, custom_prompt: dict):
        self.custom_prompt = custom_prompt
        return self

    def set_context(self, context: dict):
        self.context = context
        return self

    def set_conversation_history(self, history: list):
        self.conversation_history = history if history is not None else []
        return self

    def set_max_history_items(self, max_items: int):
        self.max_history_items = max_items
        return self

    def set_user_message(self, message: str):
        self.user_message = message
        return self

    def _replace_template_vars(self, text: str, context: dict):
        file_list = context.get("fileList", [])
        conversation_history = context.get("conversationHistory", [])

        return text.replace("{{CURRENT_PATH}}", context.get("currentPath", "/workspace")) \
                   .replace("{{FILE_COUNT}}", str(len(file_list))) \
                   .replace("{{FILE_LIST}}", json.dumps(file_list, indent=2)) \
                   .replace("{{HISTORY_COUNT}}", str(len(conversation_history))) \
                   .replace("{{CURRENT_FILE}}", context.get("currentFile", "なし")) \
                   .replace("{{CUSTOM_PROMPT_NAME}}", self.custom_prompt.get("name", "") if self.custom_prompt else "") \
                   .replace("{{OPEN_FILE_INFO}}", context.get("openFileInfo", ""))

    def build_system_prompt(self):
        if self.custom_prompt and self.custom_prompt.get("content"):
            return self.custom_prompt["content"]

        prompt = self.system_prompt
        if "contextTemplate" in SYSTEM_PROMPTS:
            prompt += "\n\n" + SYSTEM_PROMPTS["contextTemplate"]

        return self._replace_template_vars(prompt, self.context)

    def build_context_info(self):
        template = ""
        if self.context.get("openFileInfo"):
            template = CONTEXT_TEMPLATES["withOpenFile"]
        elif self.custom_prompt and self.custom_prompt.get("enabled"):
            template = CONTEXT_TEMPLATES["withCustomPrompt"]
        else:
            template = CONTEXT_TEMPLATES["basic"]

        return self._replace_template_vars(template, self.context)

    def build_conversation_history(self, provider: str):
        recent_history = self.conversation_history[-self.max_history_items:]
        messages = []

        if provider in ["claude", "openai", "local"]:
            for exchange in recent_history:
                messages.append({"role": "user", "content": exchange.get("user", "")})
                if exchange.get("ai"):
                    messages.append({"role": "assistant", "content": exchange["ai"]})
        elif provider == "gemini":
            history_text = ""
            if len(recent_history) > 0:
                for exchange in recent_history:
                    history_text += f"【ユーザー】\n{exchange.get('user', '')}\n\n"
                    if exchange.get("ai"):
                        ai_response = exchange["ai"]
                        try:
                            parsed = json.loads(ai_response)
                            history_text += f"【アシスタント】\n{parsed.get('message', ai_response)}\n\n"
                        except json.JSONDecodeError:
                            history_text += f"【アシスタント】\n{ai_response}\n\n"
            return history_text
        
        return messages

    def build_final_user_message(self):
        context_info = self.build_context_info()
        return f"{self.user_message}\n\n{context_info}"

    def build_for_provider(self, provider: str):
        system_prompt = self.build_system_prompt()

        if provider == "claude":
            claude_messages = self.build_conversation_history("claude")
            claude_messages.append({
                "role": "user",
                "content": self.build_final_user_message()
            })
            return {
                "system": system_prompt,
                "messages": claude_messages
            }
        elif provider in ["openai", "local"]:
            messages = [{"role": "system", "content": system_prompt}]
            history = self.build_conversation_history(provider)
            messages.extend(history)
            messages.append({
                "role": "user",
                "content": self.build_final_user_message()
            })
            return {"messages": messages}
        elif provider == "gemini":
            history_text = self.build_conversation_history("gemini")
            user_message = self.build_final_user_message()

            full_content = (
                "【システムプロンプト】\n" + system_prompt + "\n\n" +
                "【過去の会話】\n" + history_text +
                "【現在の質問】\n【ユーザー】\n" + user_message
            )
            return {"content": full_content}
        else:
            raise ValueError(f"Unsupported provider: {provider}")

def create_prompt_builder():
    return PromptBuilder()

def log_prompt_debug_info(prompt_data: dict, provider: str):
    if True:  # デバッグログを強制的に有効化
        print(f"--- {provider.upper()} API Request ---")
        print("Prompt Data:")
        print(json.dumps(prompt_data, indent=2, ensure_ascii=False))
        print("--------------------------")
